packages/plugin-core/src/utils/analytics.ts

```diff
@@ -94,20 +94,37 @@ export class AnalyticsUtils {
     return currentTime.minus(firstInstallTime) < ONE_WEEK;
   }
 
-  static track(event: string, props?: any) {
+  static _trackCommon({
+    event,
+    props,
+    timestamp,
+  }: {
+    event: string;
+    props?: any;
+    timestamp?: Date;
+  }) {
     const { ideVersion, ideFlavor } = AnalyticsUtils.getVSCodeIdentifyProps();
     const properties = { ...props, ...AnalyticsUtils.getCommonTrackProps() };
     const sessionId = AnalyticsUtils.getSessionId();
-    SegmentUtils.track({
+    return {
       event,
       platformProps: {
         type: AppNames.CODE,
         ideVersion,
         ideFlavor,
       },
       properties,
+      timestamp,
       integrations: { Amplitude: { session_id: sessionId } },
-    });
+    } as Parameters<typeof SegmentUtils.track>[0];
+  }
+
+  static track(event: string, props?: any) {
+    SegmentUtils.track(this._trackCommon({ event, props }));
+  }
+
+  static trackSync(opts: { event: string; props?: any; timestamp?: Date }) {
```

Comment:
That's also a bit tricky because if you do await the track call in the middle of an operation, it can block up to a few seconds. That means if we accidentally await it when not needed, we'll be introducing a significant delay for anyone with a slow connection. I think it makes sense to keep the primary function in a form where it's impossible to misuse.

Perhaps we can just do better naming? Instead of `trackSync` it could be `trackAndWait` or `trackWithPromise`.

Reply:
I think we (myself included) have a tendency to think "oh this returns a promise, so let me await it".  But I think the correct mental model should be to always evaluate "Do I need the results of the operation now?  **Only** if the answer is 'yes' should I await it. 

I think the best way to address this is with a warning in the function comments, which I had written in the underlying  track implementation (https://github.com/dendronhq/dendron/blob/20533bdf548b04d3d1c7af09bb7d4f1b520c76d7/packages/common-server/src/analytics.ts#L303), but that has since been masked with the addition of the track methods in the `SegmentUtils.ts` class. 

Another point - do we actually need a `SegmentUtils` static class when `SegmentClient` is a singleton with a static getter?  Seems redundant to me.

Reply:
fair enough. we'll also want to fix this in `common-server` though since we use this pattern there as well. will have a separate pr for that